app/workers/event_worker.rb

```diff
@@ -1,15 +1,143 @@
+  event.completed = true
+  event.completed_at = Time.now.utc
+  event.save!
+end
+
+def create_error_response(error)
+  {
+    date: '',
+    time: '',
+    status_code: '',
+    message: error.message,
+    size: '',
+    payload: {}
+  }
+end
+
+def save_http_error(event, error)
+  response = create_error_response(error)
+  event_data = JSON.parse(event.data)
+
+  event_data['outbound'].last['response'] = response
+  event.data = event_data.to_json
+  event.save
+end
+
+def create_response_object(payload, response)
+  {
+    date: DateTime.now.utc.to_s.split(' ').first,
+    time: DateTime.now.utc.to_s.split(' ')[1],
+    status_code: response.code,
+    message: response.message,
+    size: response.size,
+    payload: payload
+  }
+end
+
+def save_response(event, resp)
+  resp_code = resp.code.to_i
+  payload = (resp_code >= 300 ? {} : JSON.parse(resp.body))
+  response = create_response_object(payload, resp)
+  event_data = JSON.parse(event.data)
+
+  event_data['outbound'].last['response'] = response
+  event.data = event_data.to_json
+  event.save
+  raise Sidekiq::LargeStatusCode if resp_code >= 300
+end
+
+def create_request_object(payload, length)
+  {
+    payload: payload,
+    date: DateTime.now.utc.to_s.split(' ').first,
+    time: DateTime.now.utc.to_s.split(' ')[1],
+    content_length: length
+  }
+end
+
+def save_request(event, length, payload)
+  request = create_request_object(payload, length)
+  event_data = JSON.parse(event.data)
+
+  event_data['outbound'].push({ 'request' => request, 'response' => {} })
+  event.data = event_data.to_json
+  event.save
+end
+
+def prepend_scheme(uri)
+  return uri if uri.starts_with?('https')
+
+  minus_scheme = uri.split('//').last
+  "#{SCHEME}#{minus_scheme}"
+end
+
+def generate_http_request(bridge, payload)
+  method = bridge.method.capitalize
+  uri = URI(prepend_scheme(bridge.outbound_url))
+  http = Net::HTTP.new(uri.host, uri.port)
+  req = "Net::HTTP::#{method}".constantize.new(uri, 'Content-Type' => 'application/json')
+
+  http.use_ssl = (uri.scheme == 'https')
+  set_headers(req, bridge)
+  req.body = payload.to_json
+  [http, req]
+end
+
+def extract_payload(bridge, event)
+  if event.test
+    JSON.parse(bridge.data['test_payload'])
+  else
+    JSON.parse(bridge.data['payload'])
+  end
+end
+
+def execute_request_response_cycle(event, bridge)
+  payload = extract_payload(bridge, event)
+  http, req = generate_http_request(bridge, payload)
+
+  save_request(event, req.length, payload)
+  response = http.request(req)
+  save_response(event, response)
+end
+
 class EventWorker
   include Sidekiq::Worker
-
-  # takes event id argument
-  # find event =>
-  # find bridge =>
-  # git payload & outbound url
-  # => send it
+  sidekiq_options retry: 0
 
   def perform(event_id)
-    # do something
+    event = Event.find(event_id)
+    bridge = Bridge.find(event.bridge_id)
+
+    (bridge.retries + 1).times do
+      execute_request_response_cycle(event, bridge)
```

Comment:
This is a really creative implementation but i don't think circumventing sidekiqs retries is the best solution. Sidekiq encourages the use of it's middleware api. Here is a stackoverflow post about dynamic retries which is what we are trying to accomplish: 

https://stackoverflow.com/questions/23065578/sidekiq-retry-count-in-job

Furthermore, I don't see how you would test your custom retry implementation. Like how would you throw an error once or twice? If we use sidekiqs retry api, we don't need to test it.

Reply:
Sidekiq automatically does 25 retries over 21 days by default if you don't set retry to false or 0. We need to deactivate that if we're to let the user control the delay and retries. I tried working with setting the retry option dynamically based on bridge.retries, but I ran into issues with that so I went with doing it manually and maintaining complete control. It's a week ago so don't remember what they were exactly, but it was a pain. Right now it's dynamic in that the user sets `retries` and it's pulled from `bridge.retries`. I'll look at your SO link. 

I'm writing the events_request_spec today that will test it. If you send an http request that throws an http error, then that error will be thrown for each iteration and stored in the data object. So the error contents of the data object can be tested. Remember the `outbound` key stores an array of request-response cycles. The implementation is in the `save_http_error` and `create_error_response` methods.

Reply:
That all sounds fine. I have 1 other major concern. 

We can theoretically only run 5 jobs at one time. How are you going to do the delay function? If 5 jobs are running, and you use "sleep" or any other waiting method, those 5 jobs are stilling going. Those 5 jobs in sleep mode or whatever will hold the entire queue. If those delays are 1hr, we would have to wait 1hr until the queue can start moving again. 

This is where sidekiqs retry & delay really shine. A delay between retries won't hold up the queue but a delay during your custom retries will. How do you plan to remedy that while circumventing sidekiqs retry api?

Reply:
Merde...
Back to the drawing board